File: src/main/java/phoenix/PhoenixGroup.java
package phoenix;

import net.minecraft.block.Blocks;
import net.minecraft.item.*;
import net.minecraft.util.NonNullList;

import java.util.Comparator;

public class PhoenixGroup extends ItemGroup
{
    public PhoenixGroup(final String name)
    {
        super(name);
    }

    @Override
    public ItemStack createIcon()
    {
        return new ItemStack(Item.getItemFromBlock(Blocks.END_PORTAL_FRAME));
    }

    @Override
    public boolean hasSearchBar()
    {
        return true;
    }

    @Override
    public void fill(NonNullList<ItemStack> items)
    {
        items.sort(new comp());
        super.fill(items);
    }

    @Override
    public int getSlotColor() {
        return 0;
    }

    static class comp implements Comparator<ItemStack>
    {
        @Override
        public int compare(ItemStack i1, ItemStack i2)
        {
            int f = 0, s = 0;
            if(i1.getItem() instanceof ArmorItem)         f = 2;
            else if (i1.getItem() instanceof SwordItem)   f = 3;
            else if (i1.getItem() instanceof AxeItem)     f = 4;
            else if (i1.getItem() instanceof PickaxeItem) f = 5;
            else if (i1.getItem() instanceof ToolItem)    f = 1;
            else if (i1.getItem() instanceof BlockItem)   f = 6;

            if(i2.getItem() instanceof ArmorItem)         s = 2;
            else if (i2.getItem() instanceof SwordItem)   s = 3;
            else if (i2.getItem() instanceof AxeItem)     s = 4;
            else if (i2.getItem() instanceof PickaxeItem) s = 5;
            else if (i2.getItem() instanceof ToolItem)    s = 6;
            else if (i2.getItem() instanceof BlockItem)   s = 7;

            return Integer.compare(f, s);
        }
    }
}

File: src/main/kotlin/phoenix/PhoenixGroup.kt
package phoenix

import net.minecraft.item.*
import net.minecraft.util.IItemProvider
import net.minecraft.util.NonNullList
import phoenix.items.ash.KnifeItem

class PhoenixGroup(name: String, private val item: () -> IItemProvider) : ItemGroup(name)
{
    constructor(name: String, item: IItemProvider) : this(name, {item})

    override fun createIcon() = ItemStack(item.invoke())
    override fun hasSearchBar() = false

    override fun fill(items: NonNullList<ItemStack>)
    {
        super.fill(items)
        items.sortWith(ItemStackComparator)
    }

    private object ItemStackComparator : Comparator<ItemStack>
    {
        override fun compare(i1: ItemStack, i2: ItemStack): Int
        {
            val f = getWeight(i1)
            val s = getWeight(i2)
            return when
            {
                f > s -> 1
                f < s -> -1
                else -> i1.displayName.formattedText.compareTo(i2.displayName.formattedText)
            }
        }

        private fun getWeight(i1: ItemStack): Double
        {
            return when (i1.getItem())
            {
                is BlockItem   -> 1.0
                is SwordItem   -> 1.1
                is AxeItem     -> 1.2
                is PickaxeItem -> 1.3
                is ShovelItem  -> 1.4
                is KnifeItem   -> 1.5
                is ToolItem    -> 1.6
                is ArmorItem   -> 2.0
                else           -> 3.0
            }
        }
    }
}